simulariumio/custom_converter.py

```diff
@@ -92,3 +118,263 @@ def _read(self, input_data: CustomData) -> Dict[str, Any]:
+                local_buf[i + V1_SPATIAL_BUFFER_STRUCT.R_INDEX] = agent_data.radii[t, n]
+                n_subpoints = int(agent_data.n_subpoints[t][n])
+                if n_subpoints > 0:
+                    # add subpoints to fiber agent
+                    subpoints = [3 * n_subpoints]
+                    for p in range(n_subpoints):
+                        for d in range(3):
+                            subpoints.append(agent_data.subpoints[t][n][p][d])
+                    local_buf[
+                        i
+                        + V1_SPATIAL_BUFFER_STRUCT.NSP_INDEX : i
+                        + V1_SPATIAL_BUFFER_STRUCT.NSP_INDEX
+                        + 1
+                        + 3 * n_subpoints
+                    ] = subpoints
+                    i += (
+                        V1_SPATIAL_BUFFER_STRUCT.VALUES_PER_AGENT - 1
+                    ) + 3 * n_subpoints
+                    # optionally draw spheres at points
+                    if agent_data.draw_fiber_points:
+                        for p in range(n_subpoints):
+                            # every other fiber point
+                            if p % 2 != 0:
+                                continue
+                            # unique instance ID
+                            raw_uid = 100 * (agent_data.unique_ids[t, n] + 1) + p
+                            if raw_uid not in uids:
+                                uid = raw_uid
+                                while uid in used_unique_IDs:
+                                    uid += 100
+                                uids[raw_uid] = uid
+                                used_unique_IDs.append(uid)
+                            # add sphere
+                            local_buf[
+                                i + V1_SPATIAL_BUFFER_STRUCT.VIZ_TYPE_INDEX
+                            ] = VIZ_TYPE.DEFAULT
+                            local_buf[i + V1_SPATIAL_BUFFER_STRUCT.UID_INDEX] = uids[
+                                raw_uid
+                            ]
+                            local_buf[
+                                i + V1_SPATIAL_BUFFER_STRUCT.TID_INDEX
+                            ] = agent_data.type_ids[t, n]
+                            local_buf[
+                                i
+                                + V1_SPATIAL_BUFFER_STRUCT.POSX_INDEX : i
+                                + V1_SPATIAL_BUFFER_STRUCT.POSX_INDEX
+                                + 3
+                            ] = agent_data.subpoints[t][n][p]
+                            local_buf[i + V1_SPATIAL_BUFFER_STRUCT.R_INDEX] = 0.5
+                            i += V1_SPATIAL_BUFFER_STRUCT.VALUES_PER_AGENT - 1
+                else:
+                    i += V1_SPATIAL_BUFFER_STRUCT.VALUES_PER_AGENT - 1
+            frame_data["data"] = local_buf.tolist()
+            bundle_data.append(frame_data)
+        return bundle_data
+
+    @staticmethod
+    def _get_spatial_bundle_data_no_subpoints(
```

Comment:
when do you use this no_subpoints version?

Reply:
for some engines there's never any fibers (ReaDDy for example), so the _read_custom_data() function will use this version for that since the subpoints array will be None. I kept this because it seems like it could be a lot more performant in the case of no fibers? But haven't profiled that or anything